[Monk] Add BDB Expressions

JeremyMonk opened the pull request.

JeremyMonk committed:
[Monk] Add BDB Expressions

scamille commented:
What is bdb? Why the abbreviation, is it that long and troublesome to spell out in full?

scamille commented on engine/class_modules/monk/sc_monk.cpp:
```diff
@@ -327,16 +344,159 @@ struct monk_action_t : public Base
       p()->cooldown.bonedust_brew->adjust( timespan_t::from_seconds( time_reduction ), true );
   }
 
+   int get_bdb_zone()
```
First, this function should be const, and the ones deriving from it as well.

Second: Is it really best to return an int instead of a enum that naturally restricts the possible zones that can be returned?

scamille commented on engine/class_modules/monk/sc_monk.cpp:
```diff
@@ -327,16 +344,159 @@ struct monk_action_t : public Base
+  {
+    std::vector<player_t*> target_list = p()->sim->target_non_sleeping_list.data();
+
+    int N           = static_cast<int>( target_list.size() );
```
You seem to have very inconsistent format for local variables. Why not stick to snake_case?

scamille commented on engine/class_modules/monk/sc_monk.cpp:
```diff
@@ -327,16 +344,159 @@ struct monk_action_t : public Base
+
+    auto energy_regen_per_second = p()->resource_regen_per_second( RESOURCE_ENERGY );
+
+    // TP->SCK
```
It would be best if you can move the following 3 rdps calculations into some shared function that you call to get the rdps values. Should make your code much smaller and easier to understand.

JeremyMonk commented:
Thanks Scamille, this is derived from Tostad0ra's scholarly article here

https://colab.research.google.com/drive/1IlNnwzigBG_xa0VdXhiofvuy-mgJAhGa?usp=sharing

and I tried to implement it quickly so that an APL can be derived from it.
I will clean it up some more after work.

On Mon, Jun 6, 2022, 3:30 PM Serge Camille ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In engine/class_modules/monk/sc_monk.cpp
> <https://github.com/simulationcraft/simc/pull/6573#discussion_r890526378>:
>
> > @@ -327,16 +344,159 @@ struct monk_action_t : public Base
>        p()->cooldown.bonedust_brew->adjust( timespan_t::from_seconds( time_reduction ), true );
>    }
>
> +   int get_bdb_zone()
>
> First, this function should be const, and the ones deriving from it as
> well.
>
> Second: Is it really best to return an int instead of a enum that
> naturally restricts the possible zones that can be returned?
> ------------------------------
>
> In engine/class_modules/monk/sc_monk.cpp
> <https://github.com/simulationcraft/simc/pull/6573#discussion_r890528022>:
>
> > @@ -327,16 +344,159 @@ struct monk_action_t : public Base
>        p()->cooldown.bonedust_brew->adjust( timespan_t::from_seconds( time_reduction ), true );
>    }
>
> +   int get_bdb_zone()
> +  {
> +    std::vector<player_t*> target_list = p()->sim->target_non_sleeping_list.data();
> +
> +    int N           = static_cast<int>( target_list.size() );
>
> You seem to have very inconsistent format for local variables. Why not
> stick to snake_case?
> ------------------------------
>
> In engine/class_modules/monk/sc_monk.cpp
> <https://github.com/simulationcraft/simc/pull/6573#discussion_r890531674>:
>
> > +      // Tiger Lightning
> +      flAmp *= ( 1 + p()->spec.invoke_xuen_2->effectN( 2 ).percent() );
> +    }
> +
> +    if ( p()->buff.storm_earth_and_fire->check() )
> +    {
> +      auto sef_multiplier = ( 1 + p()->spec.storm_earth_and_fire->effectN( 1 ).percent() ) * 3;
> +      flAmp *= sef_multiplier;
> +      flAmp *= ( 2 * CO_bonus + 1 ) / 3;
> +    }
> +
> +    flAmp = flAmp * ( 1 + ( .5 * .4 * ( static_cast<double>( bdb_targets ) / N ) * ( 1 + BMH_bonus ) ) );
> +
> +    auto energy_regen_per_second = p()->resource_regen_per_second( RESOURCE_ENERGY );
> +
> +    // TP->SCK
>
> It would be best if you can move the following 3 rdps calculations into
> some shared function that you call to get the rdps values. Should make your
> code much smaller and easier to understand.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/simulationcraft/simc/pull/6573#pullrequestreview-997182105>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYEEKKATS74FELQF4UI5O2LVNZNX7ANCNFSM5X7NCZRA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JeremyMonk committed:
Clean up code

JeremyMonk committed:
Further cleaning

Almost entirely automated now this should suffice for our purposes as this likely will be deprecated with Dragonflight

JeremyMonk committed:
Fix CO and make easier to read

Hinalover committed:
[Monk] Fix errors
- Convert results into Enumerators
